Disable sharing: Filter public gallery and hide publish button

caleybrock opened the pull request:
Publish button hidden on personal project table when sharing is disabled, only for applab, gamelab, and weblab projects.

Filter applab, gamelab and weblab projects from the public gallery when the owner has sharing disabled.

davidsbailey commented on shared/middleware/helpers/auth_helpers.rb:
```diff
@@ -35,7 +35,11 @@ def get_user_sharing_disabled(user_id)
+  get_sharing_disabled_from_properties(user_properties[:properties])
+end
+
+def get_sharing_disabled_from_properties(properties)
```
in light of recent discussion on slack, we should really have a simple unit test on both of these auth helper methods too.

caleybrock commented:
Updated this logic to be on the server instead of on the client, which makes it easier to test the project_list. I also added some tests for both auth_helpers. Since @davidsbailey is dotd, @aoby could you take a look?

aoby commented on dashboard/lib/projects_list.rb:
```diff
@@ -125,6 +126,7 @@ def fetch_published_project_types(project_types, limit:, published_before: nil)
     # storage_apps and user tables. See project_and_user_fields for which
     # fields it contains.
     def get_published_project_and_user_data(project_and_user)
```
Though I know it's outside of this PR, this method description is unhelpful. Please follow the [YardDoc format](https://www.yardoc.org/) and explain what this method does (with a verb), what it expects as its `@param` and what it `@returns`. This is especially important now that it is expected to return nil in certain cases.

aoby commented on dashboard/test/lib/projects_list_test.rb:
```diff
@@ -36,4 +36,11 @@ class ProjectsListTest < ActionController::TestCase
+      properties: {sharing_disabled: true}.to_json
+    }
+    assert_nil ProjectsList.send(:get_published_project_and_user_data, project_and_user)
+  end
```
nit: "ignores" is misleading here in the test case name. I'd go with "returns nil". That's what we're testing: it returns nil for sharing_disabled projects. They are not ignored; they are inspected and rejected.

aoby commented on shared/middleware/helpers/auth_helpers.rb:
```diff
@@ -35,7 +35,12 @@ def under_13?
+def get_sharing_disabled_from_properties(properties)
+  return false unless properties
+  parsed_properties = JSON.parse(properties)
   return false unless parsed_properties
```
This line isn't needed. AFAIK `JSON.parse` will never return nil. It will either return parsed json as a hash, or raise an error:

```
irb(main):021:0> JSON.parse nil.to_json
JSON::ParserError: 784: unexpected token at 'null'

irb(main):023:0> JSON.parse nil
TypeError: no implicit conversion of nil into String

irb(main):024:0> JSON.parse nil.to_s
JSON::ParserError: A JSON text must at least contain two octets!

irb(main):026:0> JSON.parse ''
JSON::ParserError: A JSON text must at least contain two octets!

irb(main):028:0> JSON.parse '{}'
=> {}
```

aoby commented on shared/test/middleware/helpers/test_auth_helpers.rb:
```diff
@@ -0,0 +1,34 @@
+
+  def test_get_user_sharing_disabled
+    # If user is not found, default to false
+    refute get_user_sharing_disabled(nil)
```
How about the success case? Also this actually makes a call in `DASHBOARD_DB`. I'd prefer to stub that out for both a nonexistent user and a "real" user (the success case).

aoby commented on dashboard/lib/projects_list.rb:
```diff
@@ -116,7 +117,7 @@ def fetch_published_project_types(project_types, limit:, published_before: nil)
             order(Sequel.desc(:published_at)).
             limit(limit).
-            map {|project_and_user| get_published_project_and_user_data project_and_user}
+            map {|project_and_user| get_published_project_and_user_data project_and_user}.compact
```
Can we add a test for this? I'd like to verify that for a set of projects the sharing disabled ones are removed and the rest remain. I'd be ok with just 2 projects: stub the `PEGASUS_DB` call to return fake data with one sharing disabled, and the other enabled, then assert that only the enabled one remains in the result.

aoby commented:
Code LGTM. I have a few minor suggestions on improving the testing and documentation.

caleybrock commented:
Updated. Thanks for the quick feedback @aoby

aoby commented on dashboard/lib/projects_list.rb:
```diff
@@ -122,9 +122,13 @@ def fetch_published_project_types(project_types, limit:, published_before: nil)
+    # @param [object] the join of storage_apps and user tables for a published project.
+    #  See project_and_user_fields for which fields it contains.
+    # @returns [object] containing feilds relevant to the published project or
+    #  nil when the user has sharing_disabled = true
```
nit: in Ruby-land it's Hash, not object.

You can specify multiple return values
`@returns [Hash, nil]`

aoby commented on dashboard/test/lib/projects_list_test.rb:
```diff
@@ -37,10 +37,37 @@ class ProjectsListTest < ActionController::TestCase
+
+  def db_result(result)
+    stub(select: stub(join: stub(join: stub(where: stub(where: stub(exclude: stub(order: stub(limit: result))))))))
+  end
```
nit: make this private
